src/pack/scratch_variables.hpp

```diff
@@ -0,0 +1,115 @@
+//========================================================================================
+// Parthenon performance portable AMR framework
+// Copyright(C) 2025 The Parthenon collaboration
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+// (C) (or copyright) 2020-2025. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef PACK_SCRATCH_VARIABLES_HPP_
+#define PACK_SCRATCH_VARIABLES_HPP_
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <Kokkos_Core.hpp>
+
+#include "basic_types.hpp"
+#include "interface/metadata.hpp"
+#include "pack/pack_utils.hpp"
+#include "utils/type_list.hpp"
+
+namespace parthenon {
+
+inline std::string range_regex(unsigned a, unsigned b) {
+  std::ostringstream pattern;
+  pattern << "((" << std::to_string(a) << ")";
+  for (int i = a + 1; i <= b; i++) {
+    pattern << "|(" << std::to_string(i) << ")";
+  }
+  pattern << ")";
+  return pattern.str();
+}
+
+template <TopologicalType TT, int... NCOMPS>
+struct ScratchVariable : public parthenon::variable_names::base_t<true, NCOMPS...> {
+  using base_t = parthenon::variable_names::base_t<true, NCOMPS...>;
+  template <typename... Ts>
+  KOKKOS_INLINE_FUNCTION ScratchVariable(Ts &&...args)
+      : base_t(std::forward<Ts>(args)...) {}
+  static constexpr TopologicalType type = TT;
```

Comment:
I would maybe vote for `topo_type` or something instead of `type`. `T::type` makes me think it is something like 
```c++
using type = something;
```
when I read `T::type` elsewhere in the code.